src/Symfony/Component/TypeInfo/Type.php

```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
 {
     use TypeFactoryTrait;
 
-    abstract public function getBaseType(): BuiltinType|ObjectType;
-
     /**
-     * @param TypeIdentifier|class-string $subject
+     * Traverses and yields the whole type inheritance graph.
+     *
+     * For example traversing "string|\Traversable<int>" will yield
+     * - string
+     * - int
+     * - \Traversable<int>
+     * - string|\Traversable<int>
+     *
+     * @return iterable<self>
      */
-    abstract public function isA(TypeIdentifier|string $subject): bool;
+    public function traverse(): iterable
+    {
+        if ($this instanceof CompositeTypeInterface) {
+            foreach ($this->getTypes() as $type) {
+                yield from $type->traverse();
+            }
+        }
+
+        if ($this instanceof WrappingTypeInterface) {
+            yield from $this->getWrappedType()->traverse();
+        }
 
-    abstract public function asNonNullable(): self;
+        yield $this;
+    }
 
     /**
-     * @param callable(Type): bool $callable
+     * Tells if the type or one of its component (@see traverse()) is what is expected.
```

Comment:
This `isA` API is confusing IMO, because it does not allow to determine whether a type is either a subtype nor a supertype of another type.
Common questions related to types would be "does this type accept this kind of value?", which cannot be answered by this API (the best proof is the fact that both UnionType and IntersectionType uses this same implementation in the base class).
The `is` method that you removed in this PR *was* able to answer such question.

Reply:
@stof. Agreed, that is exactly one of my use case:  I need to determine whether a particular value is valid where a specific type-hint is used and this PR would render this method useless to me.
However, the current implementation has issues nevertheless: on one side `Type::object('SomeClass')->isA('SomeInterface')` but on the other `Type::union(Type::int(), Type::string())->isA(Type::int())`: it's not clear whether the type is supposed to accept what is passed to isA() or if it is the other way around. 
I was actually going to work on type comparisons, I'd suggest leaving `isA()` untouched for now and reviewing/improving the current implementation separately.

Reply:
Well, I totally agree that `isA` is weird (maybe it only needs to be renamed). But that's one more reason why removing `is` by suggesting `isA` as a replacement is a very bad idea.

Reply:
You're totally right, that's why I renamed `Type::is` to `Type::satisfy` and `Type::isA` to `Type::isIdentifiedBy` so that it is easier to understand. And I added the `CompositeTypeInterface::composedTypesSatisfy` and `WrappingTypeInterface::wrappedTypeSatisfy` so that it is easier to define whether we want to check the root, the children or the leaves.